docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+
+The event `Service` gives modules access to an event `Manager` which allows modules to emit typed and legacy,
+untyped events:
+
+```go
+package event
+type Service interface {
+    GetManager(context.Context) Manager
+}
+
+type Manager interface {
+	// Emit emits events to both clients and state machine listeners. These events MUST be emitted deterministically
+	// and should be assumed to be part of blockchain consensus.
+    Emit(proto.Message) error
+	
+	// EmitLegacy emits legacy untyped events to clients only. These events do not need to be emitted deterministically
+	// and are not part of blockchain consensus.
+    EmitLegacy(eventType string, attrs ...LegacyEventAttribute) error
+
+    // EmitClientOnly emits events only to clients. These events do not need to be emitted deterministically
+    // and are not part of blockchain consensus.
+    EmitClientOnly(proto.Message) error
+}
+```
+
+Typed events emitted with `Emit` can be observed by other modules and thus must be deterministic and should be assumed
+to be part of blockchain consensus (whether they are part of the block or app hash is left to the runtime to specify).
+
+Events emitted by `EmitLegacy` and `EmitClientOnly` are not considered to be part of consensus and cannot be observed
+by other modules. If there is a client-side need to add events in patch releases, these methods can be used.
+
+Design questions:
+* should we allow overriding event managers (like `sdk.Context.WithEventManager`)?
+
+#### Block Info Service
+
+The block info `Service` will be defined in the `cosmossdk.io/core/blockinfo` package.
+
+The `BlockInfo` service allows modules to get a limited set of information about the currently executing block
+that does not depend on any specific version of Tendermint:
+```go
+package blockinfo
+
+type Service interface {
+    GetBlockInfo(ctx context.Context) BlockInfo
+}
+
+type BlockInfo struct {
+    ChainID string
+    Height int64
+    Time time.Time
+    Hash []byte
+}
+```
+
+Only a limited set of modules need any other information from the Tendermint block header and specific versions
+of runtime modules tied to specific Tendermint versions can expose this functionality to modules that need it.
+The basic `BlockInfo` service is left fairly generic to insulate the vast majority of other modules from unneeded
+exposure to changes in the Tendermint protocol.
+
+#### Gas Service
+
+The gas `Service` and `Meter` types will be defined in the `cosmossdk.io/core/gas` package.
+
+The gas service handles tracking of gas consumptions at the block and transaction level and also allows
+for overriding the gas meter passed to child calls:
+```go
+package gas
+
+type Service interface {
+	GetMeter(context.Context) Meter
+	GetBlockMeter(context.Context) Meter
+	WithMeter(ctx context.Context, meter Meter) context.Context
+	WithBlockMeter(ctx context.Context, meter Meter) context.Context
```

Comment:
IMO `WithBlockMeter` should not be in the default Gas service. It shouldn't be exposed to most module developers. Probably only runtime and maybe antehandlers should deal with BlockGas.

Reply:
Yeah it's worth having a discussion about. IMHO there's a lot of unsafe stuff on the current sdk.Context